lib/de.ml

```diff
@@ -1382,6 +1382,302 @@ module Inf = struct
+          then
+            err_unexpected_end_of_input () ;
+          if len > bigstring_length d.o - d.o_pos
+          then
+            err_unexpected_end_of_output () ;
+          WInf.blit d.w d.i d.i_pos d.o d.o_pos len ;
+          d.o_pos <- d.o_pos + len ;
+          d.i_pos <- d.i_pos + len )
+
+    let _fill_byte d =
+      if i_rem d < 1
+      then
+        err_unexpected_end_of_input ()
+      else
+          ( if i_rem d <> 1
+            then
+              ( d.hold <- d.hold lor ((unsafe_get_uint16 d.i d.i_pos) lsl d.bits);
+                d.i_pos <- d.i_pos + 2;
+                d.bits <- d.bits + 16)
+            else
+              ( d.hold <- d.hold lor ((unsafe_get_uint8 d.i d.i_pos) lsl d.bits);
+                d.i_pos <- d.i_pos + 1;
+                d.bits <- d.bits + 8))
+    [@@inline]
+
+    let rec fill_bits d n =
+      if d.bits < n
+      then
+        ( _fill_byte d
+        ; fill_bits d n)
+
+    let pop_bits d n =
+      let v = d.hold land (1 lsl n - 1) in
+      d.hold <- d.hold lsr n ;
+      d.bits <- d.bits - n ;
+      v
+    [@@inline]
+
+    exception End
+    exception Invalid_distance
+    exception Invalid_distance_code
+
+    let inflate lit dist d =
+      try
+        let rec length () =
+          fill_bits d lit.Lookup.l ;
+          let value = lit.Lookup.t.(d.hold land lit.Lookup.m) land Lookup.mask in
+          let len = lit.Lookup.t.(d.hold land lit.Lookup.m) lsr 15 in
+          d.hold <- d.hold lsr len ;
+          d.bits <- d.bits - len ;
+          if value < 256
+          then ( unsafe_set_uint8 d.o d.o_pos value
+                ; WInf.add d.w value
+                ; d.o_pos <- d.o_pos + 1
+                ; length ())
+          else if value == 256 then raise_notrace End
+          else ( extra_length (value - 257))
+        and extra_length l =
+          let len = _extra_lbits.(l) in
+          fill_bits d len ;
+          let extra = pop_bits d len in
+          distance (_base_length.(l land 0x1f) + 3 + extra)
+        and distance l =
+          fill_bits d dist.Lookup.l ;
+          let value = dist.Lookup.t.(d.hold land dist.Lookup.m) land Lookup.mask in
+          let len = dist.Lookup.t.(d.hold land dist.Lookup.m) lsr 15 in
+          d.hold <- d.hold lsr len ;
+          d.bits <- d.bits - len ;
+          extra_distance l value;
+        and extra_distance l d_ =
+          let len = _extra_dbits.(d_ land 0x1f) in
+          fill_bits d len ;
+          let extra = pop_bits d len in
+          write l (_base_dist.(d_) + 1 + extra)
+        and write l d_ =
+          if d_ == 0 then raise_notrace Invalid_distance_code ;
+          if d_ > WInf.have d.w then raise_notrace Invalid_distance ;
+          let len = min l (bigstring_length d.o - d.o_pos) in
+          let off = WInf.mask (d.w.WInf.w - d_) in
+          if d_ == 1
+          then
+            ( let v = unsafe_get_uint8 d.w.WInf.raw off in
+              WInf.fill d.w v d.o d.o_pos len )
+          else
```

Comment:
I believe this part of the code might also need a rework, as I didnt change it at all.
The output flushing related code should still be here

Reply:
A fast look tell me that we don't need to recalculate `len` according to `(d.o_len - d.o_pos)` but we should check that `l` is lower or equal to `(d.o_len - d.o_pos)` - otherwise, you raise an error.

Then, in any case at the end of this part of the code should call `length ()` directly without any `if`.

The use of `blit`/`pre`/`rst` is ok - and I think needed.